htlcswitch/switch.go

```diff
@@ -562,37 +532,48 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
 	// payment circuit by forwarding the settle msg to the channel from
 	// which htlc add packet was initially received.
 	case *lnwire.UpdateFufillHTLC, *lnwire.UpdateFailHTLC:
-		// Exit if we can't find and remove the active circuit to
-		// continue propagating the fail over.
-		circuit, err := s.circuits.remove(packet.payHash)
-		if err != nil {
-			err := errors.Errorf("unable to remove "+
-				"circuit for payment hash: %v", packet.payHash)
-			log.Error(err)
-			return err
-		}
+		if packet.dest == (lnwire.ShortChannelID{}) {
+			// Use circuit map to find the link to forward settle/fail to.
+			circuit := s.circuits.LookupByHTLC(packet.src, packet.srcID)
+			if circuit == nil {
+				err := errors.Errorf("Unable to find source channel for HTLC "+
+					"settle/fail: channel ID = %s, HTLC ID = %d, "+
+					"payment hash = %x", packet.src, packet.srcID,
+					packet.payHash[:])
+				log.Error(err)
+				return err
+			}
 
-		// If this is failure than we need to obfuscate the error.
-		if htlc, ok := htlc.(*lnwire.UpdateFailHTLC); ok && !packet.isObfuscated {
-			htlc.Reason = circuit.ErrorEncrypter.IntermediateEncrypt(
-				htlc.Reason,
-			)
+			// Remove circuit since we are about to complete the HTLC.
+			err := s.circuits.Remove(packet.src, packet.srcID)
+			if err != nil {
+				log.Warnf("Failed to close completed onion circuit for %x: "+
+					"%s<->%s", packet.payHash[:], circuit.SrcChanID,
+					circuit.DestChanID)
+			} else {
+				log.Debugf("Closed completed onion circuit for %x: %s<->%s",
+					packet.payHash[:], circuit.SrcChanID, circuit.DestChanID)
+			}
+
+			// Obfuscate the error message for fail updates before sending back
+			// through the circuit.
+			if htlc, ok := htlc.(*lnwire.UpdateFailHTLC); ok && !packet.isObfuscated {
+				htlc.Reason = circuit.ErrorEncrypter.IntermediateEncrypt(
+					htlc.Reason)
+			}
+
+			packet.dest = circuit.SrcChanID
+			packet.destID = circuit.SrcHTLCID
```

Comment:
This is a bit hacky, why are we unable to just _always_ use the source below? (as we did before)

Reply:
I may have reversed some of the terminology because I was thinking in terms of `src` and `destination` packets, not HTLCs. Like the destination of an `update_add` is the opposite of the destination for the `update_{fail,settle}` going the the other way. Hopefully, incoming/outgoing HTLC terminology clears things up.